ulib/FStar.Matrix.fsti

```diff
@@ -0,0 +1,352 @@
+type matrix_of #c (#m #n: pos) (gen: matrix_generator c m n) = z:matrix c m n {
+  (forall (i: under m) (j: under n). ijth z i j == gen i j) /\ 
+  (forall (ij: under (m*n)). (SB.index (seq_of_matrix z) ij) == (gen (get_i m n ij) (get_j m n ij)))  
+}
+
+(* Monoid-based fold of a matrix  treated as a flat seq *)
+val foldm : (#c:Type) -> (#eq:CE.equiv c) -> (#m:pos) -> (#n:pos) -> (cm: CE.cm c eq) -> (mx:matrix c m n) -> c
+
+(* foldm_snoc of the corresponding seq is equal to foldm of the matrix *)
+val matrix_fold_equals_fold_of_seq : 
+  (#c:Type) -> (#eq:CE.equiv c) -> (#m:pos) -> (#n:pos) -> (cm: CE.cm c eq) -> (mx:matrix c m n) 
+  -> Lemma (ensures foldm cm mx `eq.eq` SP.foldm_snoc cm (seq_of_matrix mx)) [SMTPat(foldm cm mx)]
+
+(* A matrix constructed from given generator *)
+val init : (#c:Type) -> (#m:pos) -> (#n: pos) -> (generator: matrix_generator c m n) 
+  -> matrix_of generator 
+
+(* A matrix fold is equal to double foldm_snoc over init-generated seq of seqs *)
+val matrix_fold_equals_fold_of_seq_folds : (#c:Type) -> (#eq: CE.equiv c) -> 
+                                           (#m: pos) -> (#n: pos) ->
+                                           (cm: CE.cm c eq) ->
+                                           (generator: matrix_generator c m n) ->
+  Lemma (ensures foldm cm (init generator) `eq.eq`
+         SP.foldm_snoc cm (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))))
+         /\ SP.foldm_snoc cm (seq_of_matrix (init generator)) `eq.eq`
+         SP.foldm_snoc cm (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))))         
+         ) 
+
+(* This auxiliary lemma shows that the fold of the last line of a matrix
+   is equal to the corresponding fold of the generator function *)
+ 
+(* This lemma establishes that the fold of a matrix is equal to
+   nested Algebra.CommMonoid.Fold.fold over the matrix generator *)
+val matrix_fold_equals_func_double_fold  : (#c:Type) -> (#eq: CE.equiv c) -> 
+                                           (#m: pos) -> (#n: pos) ->
+                                           (cm: CE.cm c eq) ->
+                                           (generator: matrix_generator c m n) ->
+  Lemma (foldm cm (init generator) `eq.eq` 
+           CF.fold cm 0 (m-1) (fun (i:under m) -> CF.fold cm 0 (n-1) (generator i)))
+ 
+
+(* This function provides the transposed matrix generator, with indices swapped
+   Notice how the forall property of the result function is happily proved 
+   automatically by z3 :) *)
+let transposed_matrix_gen #c (#m:pos) (#n:pos) (generator: matrix_generator c m n) 
+  : (f: matrix_generator c n m { forall i j. f j i == generator i j }) 
+  = fun j i -> generator i j
+
+(* This lemma shows that the transposed matrix is 
+   a permutation of the original one *)
+let matrix_transpose_is_permutation #c (#m #n: pos) 
+                             (generator: matrix_generator c m n)
+  : Lemma (SP.is_permutation (seq_of_matrix (init generator))
```

Comment:
move the proof of the lemma to the implementation. It doesn't need to be in the .fsti

Reply:
Done